File: week_2/code/day_10/c_implement_trie.js
class TrieNode {
    constructor() {
        this.children = new Map(); // char -> TrieNode
        this.isWord = false;
    }
}

var Trie = function () {
    this.root = new TrieNode();
};

/** 
 * @param {string} word
 * @return {void}
 */
Trie.prototype.insert = function (word) {
    let curr = this.root;
    for (let i = 0; i < word.length; i++) {
        const char = word[i];
        if (!curr.children.has(char)) {
            curr.children.set(char, new TrieNode());
        }
        curr = curr.children.get(char);
    }
    curr.isWord = true;
};

/** 
 * @param {string} word
 * @return {boolean}
 */
Trie.prototype.search = function (word) {
    let curr = this.root;
    for (let i = 0; i < word.length; i++) {
        const char = word[i];
        if (!curr.children.has(char)) {
            return false;
        }
        curr = curr.children.get(char);
    }
    return curr.isWord;
};

/** 
 * @param {string} prefix
 * @return {boolean}
 */
Trie.prototype.startsWith = function (prefix) {
    let curr = this.root;
    for (let i = 0; i < prefix.length; i++) {
        const char = prefix[i];
        if (!curr.children.has(char)) {
            return false;
        }
        curr = curr.children.get(char);
    }
    return true;
};

// Example Usage
const trie = new Trie();
trie.insert("apple");
console.log(trie.search("apple"));   // return True
console.log(trie.search("app"));     // return False
console.log(trie.startsWith("app")); // return True
trie.insert("app");
console.log(trie.search("app"));     // return True

File: week_2/code/day_10/c_implement_trie.py
class TrieNode:
    def __init__(self):
        # Using a dictionary to store children nodes
        self.children = {}
        # Boolean to check if a word ends at this node
        self.is_word = False

class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        curr = self.root
        for char in word:
            if char not in curr.children:
                curr.children[char] = TrieNode()
            curr = curr.children[char]
        curr.is_word = True

    def search(self, word: str) -> bool:
        curr = self.root
        for char in word:
            if char not in curr.children:
                return False
            curr = curr.children[char]
        return curr.is_word

    def startsWith(self, prefix: str) -> bool:
        curr = self.root
        for char in prefix:
            if char not in curr.children:
                return False
            curr = curr.children[char]
        return True

# Example Usage
if __name__ == "__main__":
    trie = Trie()
    trie.insert("apple")
    print(trie.search("apple"))   # return True
    print(trie.search("app"))     # return False
    print(trie.startsWith("app")) # return True
    trie.insert("app")
    print(trie.search("app"))     # return True
